### PR TITLE
Fix use_path_as_transaction_name in Spring controllers

### PR DESCRIPTION
**Pull request by hectorespert**
## What does this PR do?

~~Fixes https://github.com/elastic/apm-agent-java/issues/838~~

Fixes `use_path_as_transaction_name` option when a transaction is generated by a Spring controller.

## Checklist

- [ ] This is an enhancement of existing features, or a new feature in existing plugins
  - [ ] I have updated [CHANGELOG.asciidoc](https://github.com/elastic/apm-agent-java/blob/master/CHANGELOG.asciidoc)
  - [ ] I have added tests that prove my fix is effective or that my feature works
  - [ ] Added an API method or config option? Document in which version this will be introduced
  - [ ] I have made corresponding changes to the documentation
- [x] This is a bugfix
  - [x] I have updated [CHANGELOG.asciidoc](https://github.com/elastic/apm-agent-java/blob/master/CHANGELOG.asciidoc)
  - [x] I have added tests that would fail without this fix
- [ ] This is a new plugin
  - [ ] I have updated [CHANGELOG.asciidoc](https://github.com/elastic/apm-agent-java/blob/master/CHANGELOG.asciidoc)
  - [ ] My code follows the [style guidelines of this project](https://github.com/elastic/apm-agent-java/blob/master/CONTRIBUTING.md#java-language-formatting-guidelines)
  - [ ] I have made corresponding changes to the documentation
  - [ ] I have added tests that prove my fix is effective or that my feature works
  - [ ] New and existing [**unit** tests](https://github.com/elastic/apm-agent-java/blob/master/CONTRIBUTING.md#testing) pass locally with my changes
  - [ ] I have updated [supported-technologies.asciidoc](https://github.com/elastic/apm-agent-java/blob/master/docs/supported-technologies.asciidoc)
  - [ ] Added an API method or config option? Document in which version this will be introduced
  - [ ] Added an instrumentation plugin? Describe how you made sure that old, non-supported versions are not instrumented by accident.
- [ ] This is something else
  - [ ] I have updated [CHANGELOG.asciidoc](https://github.com/elastic/apm-agent-java/blob/master/CHANGELOG.asciidoc)

### CODE REVIEW

**Comment by eyalkoren** [comment]
@hectorespert thanks for this PR!

#838 says:
Currently, we always prefer Spring-MVC-specific transaction name over the generic one (which is taking the `use_path_as_transaction_name` config into consideration). We want to fix that so that if the method name is known, **we still do the same**, but if the method name is unknown, we should prefer the default/generic transaction name setting.
The way to do that is by using different priorities - currently, we always use `PRIO_HIGH_LEVEL_FRAMEWORK` when setting transaction name through the Spring MVC configurations. What I think @felixbarny meant is that we will use this priority only if the method name is known, otherwise we should use `PRIO_LOW_LEVEL_FRAMEWORK + 1`, allowing the default name-setting to take effect, including the `use_path_as_transaction_name` config.

So, while this PR may be considered as a "fix" for the current behaviour, it is not doing what #838 aims for and it defies our current assumption that framework-specific transaction names are always preferable. 
So, while the validity of the current assumption is debatable, we wouldn't want to change it atm.

**Comment by hectorespert** [comment]
@eyalkoren I changed the scope of this PR. It's not a fix for https://github.com/elastic/apm-agent-java/issues/838 by I link that because I started trying to fit it.

**Comment by eyalkoren** [comment]
@hectorespert I see.  As I mentioned above- we give high-level framework naming higher priority comparing to the `use_path_as_transaction_name` config, and we wouldn't want to change it  at them moment. 
So, again, while the validity of this decision is debatable, let's not do it here. If we _do_ decide to change it, it wouldn't be a Spring-MVC-specific change, it should be consistent with all the other frameworks.

Please let me know if and how you want to proceed, or whether you want me to take over.
Thanks!

**Comment by hectorespert** [comment]
> Please let me know if and how you want to proceed, or whether you want me to take over.

@eyalkoren There is no problem if you take over.

File: apm-agent-plugins/apm-spring-webmvc-plugin/src/test/java/co/elastic/apm/agent/springwebmvc/SpringTransactionNameInstrumentationTest.java [hunk 1]
```diff
@@ -79,13 +85,46 @@ void testControllerTransactionName() throws Exception {
         assertThat(reporter.getFirstTransaction().getNameAsString()).isEqualTo("TestController#test");
     }
 
+    @Test
+    void testControllerTransactionNameWhenUsePathAsTransactionNameIsTrue() throws Exception {
+        WebConfiguration webConfiguration = config.getConfig(WebConfiguration.class);
+        when(webConfiguration.isUsePathAsName()).thenReturn(true);
+        mockMvc.perform(get("/test/944207fa-d27b-40f6-be58-6dd6a1b541d4"))
+            .andExpect(content().string("944207fa-d27b-40f6-be58-6dd6a1b541d4"));
+        assertThat(reporter.getFirstTransaction().getNameAsString()).isEqualTo("GET /test/944207fa-d27b-40f6-be58-6dd6a1b541d4");
```
Comment (SylvainJuge) [inline apm-agent-plugins/apm-spring-webmvc-plugin/src/test/java/co/elastic/apm/agent/springwebmvc/SpringTransactionNameInstrumentationTest.java:94]: We need to minimize the name cardinality, in this case we should expect to have `"/test/{uuid}"` as transaction name, not the one that contains the full path. The full URL path is still captured in transaction attributes so it's not lost and can be retrieved if needed.

**Comment by eyalkoren** [comment]
Superseded by #1906 as the fix for #838.

While this PR offers additional value, the full implementation for Spring MVC should be similar to what we have for JAX-RS, where we properly read annotation values to infer dynamic path parts and account for annotation inheritance, and this is not something we want to pursue at this point.

Thanks for raising this to awareness @hectorespert 🙏

**Pull request closed without merging**